### PR TITLE
Optimize codebase

### PR DESCRIPTION
**Pull request by madcake**

### CODE REVIEW

File: services/controllers/charts/base.go [hunk 1]
```diff
@@ -158,3 +157,24 @@ func (c Controller) getChartsFromMemory(coin uint, token string) ([]models.Ticke
 	}
 	return []models.Ticker{result}, nil
 }
+
+func cutChartByMaxItems(chart watchmarket.Chart, maxItems int) watchmarket.Chart {
```
Comment (hewigovens) [inline services/controllers/charts/base.go:161]: add some tests?

**Comment by hewigovens** [comment]
CI was failed

File: services/controllers/models.go [hunk 2]
```diff
@@ -64,3 +76,38 @@ type (
...
+	if currency == "" {
+		currency = watchmarket.DefaultCurrency
+	}
+	return currency
+}
+
+func GetTimeStart(rawTime string) int64 {
+	timeStart, err := strconv.ParseInt(rawTime, 10, 64)
+	if err != nil {
+		timeStart = time.Now().Unix() - 60*60*24
```
Comment (vikmeup) [inline services/controllers/models.go:102]:
```suggestion
		timeStart = time.Now().Unix() - time.Hour * 24
```

File: services/controllers/models.go [hunk 3]
```diff
@@ -64,3 +76,38 @@ type (
...
 	}
 )
+
+func GetCoinId(rawCoinId string) (uint, error) {
+	coinId, err := strconv.Atoi(rawCoinId)
+	if err != nil {
+		return 0, err
+	}
+	if _, ok := coin.Coins[uint(coinId)]; !ok {
+		return 0, errors.New(watchmarket.ErrBadRequest)
```
Comment (vikmeup) [inline services/controllers/models.go:86]:
```suggestion
		return 0, errors.New("invalid coin Id: " + coinId)
```

This is generic function, we should keep error separate from http error messages

File: services/controllers/charts/base.go (hunk 1 above)
Comment (vikmeup) [inline services/controllers/charts/base.go:161]:
```suggestion
func calculateChartByMaxItems(chart watchmarket.Chart, maxItems int) watchmarket.Chart {
```

File: services/controllers/charts/base.go [hunk 4]
```diff
@@ -96,21 +95,21 @@ func (c Controller) hasTickers(coin uint, token string) bool {
 	return len(tickers) > 0
 }
 
-func (c Controller) getChartsFromApi(data chartsNormalizedRequest) (ch watchmarket.Chart, err error) {
+func (c Controller) getChartsFromApi(data controllers.ChartRequest) (ch watchmarket.Chart, err error) {
```
Comment (vikmeup) [inline services/controllers/charts/base.go:98]: 👍

File: services/controllers/charts/base.go [hunk 5]
```diff
@@ -69,21 +68,21 @@ func (c Controller) HandleChartsRequest(cr controllers.ChartRequest) (ch watchma
...
 }
 
-func (c Controller) hasTickers(coin uint, token string) bool {
+func (c Controller) hasTickers(coinId uint, tokenId string) bool {
 	var tickers []models.Ticker
 	var err error
 
 	if c.useMemoryCache {
-		if tickers, err = c.getChartsFromMemory(coin, token); err != nil {
+		if tickers, err = c.getChartsFromMemory(coinId, tokenId); err != nil {
```
Comment (vikmeup) [inline services/controllers/charts/base.go:81]: One suggestion to use `asset` as one param, instead of `(coinId, tokenId)`

**Review by ivanKis — COMMENTED** [review]
Checked requests for  watch market url: https://market-codebase-clean-c-ulhyxo.herokuapp.com/
No regression was found.

**Review by vikmeup — APPROVED** [review]
I haven't looked into details, but should be good it works as expected

**Pull request merged**